geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/collection/IntBitSet.scala

```diff
@@ -0,0 +1,508 @@
+/***********************************************************************
+ * Copyright (c) 2013-2019 Commonwealth Computer Research, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Apache License, Version 2.0
+ * which accompanies this distribution and is available at
+ * http://www.opensource.org/licenses/apache2.0.php.
+ ***********************************************************************/
+
+package org.locationtech.geomesa.utils.collection
+
+import com.esotericsoftware.kryo.io.{Input, Output}
+
+/**
+  * A bit set backed by ints (to reduce memory footprint for small sizes).
+  *
+  * Provides access to the underlying mask, for serialization. Not thread safe.
+  */
+sealed abstract class IntBitSet {
+
+  /**
+    * Checks whether the value is contained in the set or not
+    *
+    * @param value value to check
+    * @return true if contains, false otherwise
+    */
+  def contains(value: Int): Boolean
+
+  /**
+    * Adds the value to the set, if it is not present. Note that implementations may be constructed with
+    * a fixed capacity - attempting to add a value outside the capacity will result in undefined behavior
+    *
+    * @param value value to add
+    * @return true if value was added, false if value was already present
+    */
+  def add(value: Int): Boolean
+
+  /**
+    * Removes the value from the set, if it is present
+    *
+    * @param value value to remove
+    * @return true if value was removed, false if value was not present
+    */
+  def remove(value: Int): Boolean
+
+  /**
+    * Resets all values
+    */
+  def clear(): Unit
+
+  /**
+    * Serialize to a kryo output stream
+    *
+    * @param output output
+    */
+  def serialize(output: Output): Unit
+}
+
+object IntBitSet {
```

Comment:
This is an improvement over other BitSets in that it can be very small?

Reply:
it just minimizes the serialization footprint - normally bitsets store values in a Long, so 64 values and 8 bytes per 'word'. using ints instead means 32 values and 4 bytes per word, so we can save some bytes if we have between 0-32 values, or 64-96, etc.